DSpace/dspace-angular
# Hard redirect after log in

## References
* Fixes #635
* Fixes #756
* Fixes #712 

## Description
This PR ensures a hard redirect is triggered after the user logs in. The purpose of doing so is to ensure everything in the ngrx store was retrieved using the correct authorizations.

After logging in and the hard redirect is triggered, the user is redirected back to the page they were originally on (before being sent to /login, or the page they opened the dropdown on).

This fixes several cache and authentication related issues.

## Instructions for Reviewers
Changes made:
- Modified `AuthenticatedGuard` to return a UrlTree to `/login` when the user is not authenticated. This used to contain a dispatch of a `RedirectWhenAuthenticationIsRequiredAction`, which caused the guard to finish async to other guards. Important to note is that this guard only emits a value when loading in the AuthState is false.
- Added new `ReloadGuard` to the `/reload` route, which now accepts a `redirect` url parameter and will redirect the user to the redirect url when present, or `/home` when absent. (the use of this is explained below)
- Added new service `HardRedirectService` and two implementations for both the browser and server app module. These services contain a `redirect()` method, which performs a hard redirect to the provided url
- Modified AuthService's `navigateToRedirectUrl` to call `HardRedirectService.redirect()` to `/reload` with the url as the `redirect` query parameter
- The redirect URL in the store is set in two possible ways:
-- The user visits a page and is redirected to /login by the authenticated-guard. The guard will set the redirect URL to the route the user originally tried to go to.
-- The user visits a page and submits a login using the dropdown at the top of the page. The submit methods (in password and shiboleth components) will set the redirect URL to the current page, before triggering the login.
- Another important change to ensure the app is only loaded when the authenticated and hard redirect is finished, is the addition of the `blocking` property in the AuthState
-- This property is set to "true" by default and is only set to "false" when the redirect action happens
-- The entire app (app-component) is wrapped in an `*ngIf` that is true when blocking is false, [as suggested on #756](https://github.com/DSpace/dspace-angular/issues/756#issuecomment-663054226)
-- On top of the `*ngIf`, all routes are locked behind the new `AuthBlockingGuard`, which returns true once blocking turns from true to false. This is because we noticed angular was still loading resolvers for parts of the template inside the false ngIf. Presumably this is a performance optimisation from the framework.

How to test:
(Whenever I mention a "hard redirect", the page should perform a hard redirect, navigate to `/reload/...`, display a loading indicator in the middle of your screen and eventually redirect you back to your original page)
- Visit a route locked behind by AuthenticatedGuard as anonymous. This should redirect you to /login. After logging in, a hard redirect should take place, ending up on the page you originally tried to access.
- Navigate to a random public page as anonymous and login using the dropdown on top of the page. A hard redirect should take place, ending up on the page you were originally on.
- Manually visit `/login` and login. A hard redirect should take place, ending up on the homepage.
- Content locked behind authentication should display correctly right after the hard redirect takes place. See #635 for more details.
- Every rest request after the redirect should contain the authorization header. See #756 for more details.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
